File: Kattis/CrackingTheCode/code.cpp
#include <iostream>
#include <map>
#include <vector>
using namespace std;

int main() {
  int num_tests;
  cin >> num_tests;
  for (int t = 0; t < num_tests; t++) {

    int num_encrypted_messages;
    cin >> num_encrypted_messages;
    vector<string> messages(num_encrypted_messages);
    for (int i = 0; i < num_encrypted_messages; i++) {
      cin >> messages[i];
    }
    string decrypted_message;
    cin >> decrypted_message;
    string message_to_decrypt;
    cin >> message_to_decrypt;
    vector<string> decryption_results(num_encrypted_messages);

    for (int i = 0; i < num_encrypted_messages; i++) {
      map<char, char> encrypt_to_decrypt;
      map<char, char> decrypt_to_encrypt;

      if (messages[i].length() != decrypted_message.length()) {
        decryption_results[i] = "\0";
        continue;
      }

      bool invalid = false;
      for (int j = 0; j < messages[i].length(); j++) {
        char e = messages[i].at(j);
        char d = decrypted_message.at(j);
        if (encrypt_to_decrypt.count(e) != 0 && encrypt_to_decrypt[e] != d) {
          invalid = true;
        } else {
          encrypt_to_decrypt[e] = d;
        }
        if (decrypt_to_encrypt.count(d) != 0 && decrypt_to_encrypt[d] != e) {
          invalid = true;
        } else {
          decrypt_to_encrypt[d] = e;
        }
      }


      if (invalid) {
        decryption_results[i] = "\0";
        continue;
      }

      string str = "";
      for (int j = 0; j < message_to_decrypt.length(); j++) {
        if (encrypt_to_decrypt.count(message_to_decrypt[j]) != 0) {
          str += (encrypt_to_decrypt[message_to_decrypt[j]]);
        } else {
          str += '?';
        }
      }

      decryption_results[i] = str;
    }

    string result = "";
    for (int i = 0; i < num_encrypted_messages; i++) {
      if (decryption_results[i].length() != 0) {
        result = decryption_results[i];
        break;
      }
    }

    if (result.length() == 0) {
      cout << "IMPOSSIBLE" << endl;
      continue;
    }

    for (int i = 0; i < num_encrypted_messages; i++) {
      if (decryption_results[i].length() == 0) continue;
      for (int j = 0; j < decryption_results[i].length(); j++) {
        if (result[j] == '?' || decryption_results[i].at(j) == '?') {
          result[j] = '?';
        } else if (result[j] != decryption_results[i].at(j)) {
          result[j] = '?';
        }
      }
    }

    int missing = 0;
    for (int i = 'a'; i <= 'z'; i++) {
      if (result.find(i) == -1) missing++;
    }

    if (missing == 1) {
      for (int i = 'a'; i <= 'z'; i++) {
        if (result.find(i) == -1) {
          result[result.find('?')] = (char)i;
        }
      }
    }

    cout << result << endl;

  }
  return 0;
}

File: Kattis/CrackingTheCode/Code.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Code {
  public static void main(String[] args) throws Exception {
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    int numTests = Integer.parseInt(in.readLine());
    for (int t = 0; t < numTests; t++) {

      int numEncryptedMessages = Integer.parseInt(in.readLine());
      String[] messages = new String[numEncryptedMessages];
      for (int i = 0; i < numEncryptedMessages; i++) {
        messages[i] = in.readLine();
      }
      String decryptedMessage = in.readLine();
      String messageToDecrypt = in.readLine();
      String[] decryptionResults = new String[messages.length];

      for (int i = 0; i < messages.length; i++) {
        HashMap<Character, Character> encryptToDecrypt = new HashMap<Character,
            Character>();
        HashMap<Character, Character> decryptToEncrypt = new HashMap<Character,
            Character>();

        if (messages[i].length() != decryptedMessage.length()) {
          decryptionResults[i] = null;
          continue;
        }

        boolean invalid = false;
        for (int j = 0; j < messages[i].length(); j++) {
          char e = messages[i].charAt(j);
          char d = decryptedMessage.charAt(j);
          if (encryptToDecrypt.get(e) != null && encryptToDecrypt.get(e) != d) {
            invalid = true;
            // System.out.println("Broke mapping e:"+e+" to d:"+d);
          } else {
            encryptToDecrypt.put(e, d);
          }
          if (decryptToEncrypt.get(d) != null && decryptToEncrypt.get(d) != e) {
            // System.out.println("Broke mapping e:"+e+" to d:"+d+" in decrypt to encrypt");
            invalid = true;
          } else {
            decryptToEncrypt.put(d, e);
          }
        }

        if (invalid) {
          decryptionResults[i] = null;
          continue;
        }

        StringBuilder sb = new StringBuilder();
        for (int j = 0; j < messageToDecrypt.length(); j++) {
          if (encryptToDecrypt.get(messageToDecrypt.charAt(j)) != null) {
            sb.append(encryptToDecrypt.get(messageToDecrypt.charAt(j)));
          } else {
            sb.append("?");
          }
        }

        decryptionResults[i] = sb.toString();
      }

      String result = null;
      for (String s : decryptionResults) {
        if (s != null) {
          result = s;
          break;
        }
      }

      if (result == null) {
        System.out.println("IMPOSSIBLE");
        continue;
      }

      char[] resultArr = result.toCharArray();
      for (String s : decryptionResults) {
        if (s == null) continue;
        for (int i = 0; i < s.length(); i++) {
          if (resultArr[i] == '?' || s.charAt(i) == '?') {
            resultArr[i] = '?';
          } else if (resultArr[i] != s.charAt(i)) {
            resultArr[i] = '?';
          }
        }
      }

      System.out.println(new String(resultArr));

    }
  }
}